executor/analyze.go

```diff
@@ -731,12 +819,68 @@ func (e *AnalyzeColumnsExec) buildSamplingStats(ranges []*ranger.Range, needExtS
 		e.job.Update(subCollector.Count)
 		rootRowCollector.MergeCollector(subCollector)
 	}
+
+	// handling virtual columns
+	var hasVirtualCol bool
+	var chk *chunk.Chunk
+	fieldTps := make([]*types.FieldType, 0, len(schema.Columns))
+	virtualColIdx := make([]int, 0)
+	for i, col := range schema.Columns {
+		fieldTps = append(fieldTps, col.RetType)
+		if col.VirtualExpr != nil {
+			hasVirtualCol = true
+			virtualColIdx = append(virtualColIdx, i)
+		}
+	}
+	needNormalDecode := true
+	if hasVirtualCol {
+		// decode + eval virtual columns
+		chk = chunk.NewChunkWithCapacity(fieldTps, len(rootRowCollector.Samples))
+		decoder := codec.NewDecoder(chk, e.ctx.GetSessionVars().TimeZone)
+		for _, sample := range rootRowCollector.Samples {
+			for i := range sample.Columns {
+				_, err := decoder.DecodeOne(sample.Columns[i].GetBytes(), i, fieldTps[i])
+				if err != nil {
+					// Eval virtual column failed. Give up building virtual column stats.
+					logutil.BgLogger().Warn("Build stats on virtual column failed when analyzing. Stats on virtual column will be incomplete",
+						zap.Int64("table id", e.AnalyzeInfo.TableID.TableID),
+						zap.Int64("partition id", e.AnalyzeInfo.TableID.PartitionID),
+						zap.Error(err))
+					goto NORMALDECODE
+				}
+			}
+		}
+		err := FillVirtualColumnValue(fieldTps, virtualColIdx, schema, e.colsInfo, e.ctx, chk)
+		if err != nil {
+			// Eval virtual column failed. Give up building virtual column stats.
+			logutil.BgLogger().Warn("Build stats on virtual column failed when analyzing. Stats on virtual column will be incomplete",
+				zap.Int64("table id", e.AnalyzeInfo.TableID.TableID),
+				zap.Int64("partition id", e.AnalyzeInfo.TableID.PartitionID),
+				zap.Error(err))
+			goto NORMALDECODE
+		}
+		iter := chunk.NewIterator4Chunk(chk)
+		for row, i := iter.Begin(), 0; row != iter.End(); row, i = iter.Next(), i+1 {
+			datums := row.GetDatumRow(fieldTps)
+			rootRowCollector.Samples[i].Columns = datums
+		}
+		needNormalDecode = false
+	}
+NORMALDECODE:
+	if needNormalDecode {
+		// normal decode
+		for _, sample := range rootRowCollector.Samples {
+			for i := range sample.Columns {
+				sample.Columns[i], err = tablecodec.DecodeColumnValue(sample.Columns[i].GetBytes(), &e.colsInfo[i].FieldType, sc.TimeZone)
+				if err != nil {
+					return 0, nil, nil, nil, nil, err
+				}
+			}
+		}
+	}
+
```

Comment:
Can we combine these 2 loops for decoding into a single one?

Reply:
Use a new function to wrap the upper one.